accounts/abi/abi_test.go

```diff
@@ -29,6 +29,180 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
+// formatSilceOutput add padding to the value and adds a size
+func formatSliceOutput(v ...[]byte) []byte {
+	off := common.LeftPadBytes(big.NewInt(int64(len(v))).Bytes(), 32)
+	output := append(off, make([]byte, 0, len(v)*32)...)
+
+	for _, value := range v {
+		output = append(output, common.LeftPadBytes(value, 32)...)
+	}
+	return output
+}
+
+// quick helper padding
+func pad(input []byte, size int, left bool) []byte {
+	if left {
+		return common.LeftPadBytes(input, size)
+	}
+	return common.RightPadBytes(input, size)
+}
+
+func TestTypeCheck(t *testing.T) {
+	for i, test := range []struct {
+		typ   string
+		input interface{}
+		err   error
+	}{
+		{"uint", big.NewInt(1), nil},
+		{"int", big.NewInt(1), nil},
+		{"uint30", big.NewInt(1), nil},
+		{"uint30", uint8(1), varErr(reflect.Ptr, reflect.Uint8)},
+		{"uint16", uint16(1), nil},
+		{"uint16", uint8(1), varErr(reflect.Uint16, reflect.Uint8)},
+		{"uint16[]", []uint16{1, 2, 3}, nil},
+		{"uint16[]", [3]uint16{1, 2, 3}, nil},
+		{"uint16[]", []uint32{1, 2, 3}, typeErr(formatSliceString(reflect.Uint16, -1), formatSliceString(reflect.Uint32, -1))},
+		{"uint16[3]", [3]uint32{1, 2, 3}, typeErr(formatSliceString(reflect.Uint16, 3), formatSliceString(reflect.Uint32, 3))},
+		{"uint16[3]", [4]uint16{1, 2, 3}, typeErr(formatSliceString(reflect.Uint16, 3), formatSliceString(reflect.Uint16, 4))},
+		{"uint16[3]", []uint16{1, 2, 3}, nil},
+		{"uint16[3]", []uint16{1, 2, 3, 4}, typeErr(formatSliceString(reflect.Uint16, 3), formatSliceString(reflect.Uint16, 4))},
+		{"address[]", []common.Address{common.Address{1}}, nil},
+		{"address[1]", []common.Address{common.Address{1}}, nil},
+		{"address[1]", [1]common.Address{common.Address{1}}, nil},
+		{"address[2]", [1]common.Address{common.Address{1}}, typeErr(formatSliceString(reflect.Array, 2), formatSliceString(reflect.Array, 1))},
+		{"bytes32", [32]byte{}, nil},
+		{"bytes32", [33]byte{}, typeErr(formatSliceString(reflect.Uint8, 32), formatSliceString(reflect.Uint8, 33))},
+		{"bytes32", common.Hash{1}, nil},
+		{"bytes31", [31]byte{}, nil},
+		{"bytes31", [32]byte{}, typeErr(formatSliceString(reflect.Uint8, 31), formatSliceString(reflect.Uint8, 32))},
```

Comment:
It would be easier to see the expected behaviour if you just put the error string here.

Compare

```go
		{"bytes31", [32]byte{}, typeErr(formatSliceString(reflect.Uint8, 31), formatSliceString(reflect.Uint8, 32))},
```

with 

```go
		{"bytes31", [32]byte{}, "abi: cannot use [32]byte as type [31]byte as argument"},
```